### DISCUSSION TITLE
Neat Owlkettle Project ideas

### DISCUSSION THREAD

**Discussion by PhilippMDoerner** [can-lehmann/owlkettle#147] [General]
Honestly this is just me spitballing some ideas around to consider for myself later when owlkettle has more Widgets wrapped.
These could also be neat project ideas to point somebody to if they want to try out owlkettle but can't think of anything to build with it.

I'll expand the list here as I come up with more ideas or we can generally post more suggestions. 

Here's the list:
1) GUI Representation of nim docs
2) GUI for atlas

---

### 1) GUI representation of nim docs
Write an application that reads in JSON output of `nim jsondoc` and visualizes it, similar to the HTML output. 
The application would be somewhat analogous to GTK's DevHelp application.

Easy and fast search for symbols, and their explanations, representation of modules etc. 

Bonus feature down the line could be to enable opening a nim project in said application and it generates the JSON of nim jsondoc itself.

---

### 2) GUI for atlas
Nim 2.0 comes with the first iteration of a replacement for nimble: Atlas.
Could be neat to write a small GUI to use atlas.

Features it could have in general:
- Creating a workspace
- Creating a project in a workspace
- Launching a custom command when "entering" a project to open it in an IDE (e.g. clicking on a project runs a command that is user defined to open it in vscode, nvim, emacs, intellij or the like for you)

Features when looking at a specific project:
- Read in the nimble file and visualize the available meta-data: Dependencies, their version, the author, the license, description, the project version
- Provide actions to upgrade the project version
- Show in the list of dependencies if there are new versions that a given dependency could be upgraded to (and provide buttons to upgrade said version)
- Provide a command to add a dependency to a project (including installing it and adding it to the nimble file)
- Provide a command to remove a dependency from a project (including removing it from the nimble file)
- Provide buttons to execute a nimble task of the project

I'm sure there's more that could be done by just looking at the features the atlas CLI provides with `atlas -h` and seeing what could be sensibly wrapped and represented in a GUI.

---

Note: In my eyes both of these currently face a blocker because in my opinion I'd want to have GtkStack fully wrapped and usable with GtkStackSidebar, which means we need to have a solved #115 as GtkStackSidebar requires a reference to the same stack that is being rendered.

**Comment by can-lehmann** [comment]
Building nice sidebars & applications with multiple pages is entirely possible without using GtkStack or GtkStackSidebar. You can use a `case` statement instead of a GtkStack and a `ListBox` with `style = [ListBoxNavigationSidebar]` instead of the StackSidebar (like [here](https://github.com/can-lehmann/owlkettle/blob/main/examples/widgets/adw/flap.nim#L55)).

**Reply by PhilippMDoerner** [reply]
While true, the animations of GtkStack are cool enough that I'd want to use them just because it'd "feel" more fun ^^

**Comment by can-lehmann** [comment]
Generally though, I think it would be an important step for owlkettle to build an app and get it published on https://flathub.org/. I think it makes sense to build something that is useful not just to people using Nim. Maybe something like a nice vector graphics editor. The aim is building an app that people actually want to use and not an app that is just "the one owlkettle app on flathub".  Ideally we could also use that experience to write a guide on publishing apps for the docs afterwards.

**Reply by can-lehmann** [reply]
It might also be an option to continue the development of https://github.com/can-lehmann/Graphing and publish that.
